parser.go

```diff
@@ -0,0 +1,208 @@
+package main
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"os"
+	"sort"
+	"strings"
+)
+
+// Parser Structure: dictionary, sections
+// It's basically a map of a map, map of sections where each section maps to keys : values
+// map[section] --> returns a map where each key --> value
+type Parser struct {
+	dictionary map[string]map[string]string
+}
+
+// NewParser creates an instance of our Parser, which is iniitially an empty map of maps
+func NewParser() Parser {
+	return Parser{make(map[string]map[string]string)}
+}
+
+// ErrCouldNotOpen happens when file cannot be opened and provides file name
+var ErrCouldNotOpen error
+
+// ErrMissingValueAssignment happens when a key isn't followed by an = statement
+var ErrMissingValueAssignment error
+
+// ErrSectionNameMissingClosure happens when section name is missing the ] paranthesis
+var ErrSectionNameMissingClosure error
+
+// ErrWrongParanthesisOrder happens when section name starts by the wrong paranthesis ']'
+var ErrWrongParanthesisOrder error
+
+// ErrInvalidSectionName happens when section is written in a wrong form --> ex: sectionName]
+var ErrInvalidSectionName error
+
+// ErrCouldNotWriteToFile happens when file cannot be written to
+var ErrCouldNotWriteToFile error
+
+// LoadFromFile loads ini file
+// Saves all lines locally into an array of strings
+func (parsedMap *Parser) LoadFromFile(fileName string) error {
+	var iniLines []string
+	var input io.Reader
+
+	file, err := os.Open(fileName)
+	if err != nil {
+		ErrCouldNotOpen = fmt.Errorf("cannot open file: %s", fileName)
+		return ErrCouldNotOpen
+	}
+	defer file.Close()
+
+	input = file
+	scanner := bufio.NewScanner(input)
+	for scanner.Scan() {
+		iniLines = append(iniLines, scanner.Text())
+	}
+	return parsedMap.parserLogic(iniLines)
+}
+
+// LoadFromString loads ini script from a string
+// Saves all lines locally into an array of strings
+func (parsedMap *Parser) LoadFromString(str string) error {
+	iniLines := strings.Split(str, "\n")
+	return parsedMap.parserLogic(iniLines)
+}
+func (parsedMap *Parser) initializer() {
+	if parsedMap.dictionary == nil {
+		parsedMap.dictionary = make(map[string]map[string]string)
+	}
+}
```

Comment:
why initializer? you already have `NewParser`

Reply:
> why initializer? you already have `NewParser`

well, sorry turns out this step is already unnecessary because it's the external map which gets initialized while creating Parser struct by default
I removed it from code and it didn't cause any errors

Reply:
any initialization (in this case calling NewParser) should be the responsibility of the user so don't bother checking if nil anyways

Reply:
> any initialization (in this case calling NewParser) should be the responsibility of the user so don't bother checking if nil anyways

aha got it, tmaam